core/src/main/java/org/apache/iceberg/encryption/EnvelopeEncryptionManager.java

```diff
@@ -0,0 +1,146 @@
+    this.kmsClient = Preconditions.checkNotNull(kmsClient,
+        "Cannot create EnvelopeEncryptionManager because KmsClient is null");
+    this.pushdown = pushdown;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    EnvelopeMetadata metadata = generateEnvelopeMetadata(dataEncryptionConfig);
+
+    // Pushdown to data formats with native encryption support (such as Parquet)
+    NativeFileCryptoParameters nativeEncryptParams = NativeFileCryptoParameters.create(metadata.dek())
+        .encryptionAlgorithm(metadata.algorithm().toString())
+        .build();
+
+    try {
+      ((NativelyEncryptedFile) rawOutput).setNativeCryptoParameters(nativeEncryptParams);
+    } catch (ClassCastException e) {
+      throw new RuntimeException("Can't natively encrypt " + rawOutput.location() + " because the class " +
+          rawOutput.getClass() + " doesn't implement NativelyEncryptedFile interface", e);
+    }
+
+    return new BaseEncryptedOutputFile(rawOutput, metadata);
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    EnvelopeMetadata metadata = EnvelopeMetadataParser.fromJson(encrypted.keyMetadata().buffer());
+    ByteBuffer fileDek = kmsClient.unwrapKey(metadata.wrappedDek(), metadata.kekId());
+
+    // Pushdown to data formats with native decryption support (such as Parquet)
+    NativeFileCryptoParameters nativeDecryptParams = NativeFileCryptoParameters.create(fileDek).build();
+    InputFile rawInput = encrypted.encryptedInputFile();
+
+    try {
+      ((NativelyEncryptedFile) rawInput).setNativeCryptoParameters(nativeDecryptParams);
+    } catch (ClassCastException e) {
+      throw new RuntimeException("Can't natively decrypt " + rawInput.location() + " because the class " +
+          rawInput.getClass() + " doesn't implement NativelyEncryptedFile interface", e);
+    }
+
+    return rawInput;
+  }
+
+  private EnvelopeMetadata generateEnvelopeMetadata(EnvelopeConfig config) {
+    if (null == workerRNG) {
+      workerRNG = new SecureRandom();
+    }
+
+    String tableKekID = config.kekId();
+    byte[] fileDek;
+    String wrappedFileDEK;
+
+    if (kmsClient.supportsKeyGeneration()) {
+      KmsClient.KeyGenerationResult generatedDek = kmsClient.generateKey(tableKekID);
+      fileDek = generatedDek.key().array();
+      wrappedFileDEK = generatedDek.wrappedKey();
+    } else {
+      fileDek = new byte[dataKeyLength];
+      workerRNG.nextBytes(fileDek);
+      wrappedFileDEK = kmsClient.wrapKey(ByteBuffer.wrap(fileDek), tableKekID);
+    }
+
+    EnvelopeMetadata fileEnvelopeMetadata = new EnvelopeMetadata(tableKekID, wrappedFileDEK, null,
+            config.algorithm());
+    fileEnvelopeMetadata.setDek(ByteBuffer.wrap(fileDek));
+
+    return fileEnvelopeMetadata;
+  }
+
+  public static KmsClient loadKmsClient(String classPath, Map<String, String> properties) {
```

Comment:
Are we going to have a dedicated Kms utility class in the future? This method should fit better there. I'm OK with the current solution though, if EnvelopeEncryptionManager is the only place to use KMS client.

Reply:
Yep, I know.. Currently, no Kms utilities (besides this function) are planned.

Reply:
Maybe we can start to create one KMS utility.

Reply:
Sure, why not. Will add.